### PR TITLE
feat: initial implementation of key value block attributes

### PR DESCRIPTION
**Pull request by tachibanayui**
This is my initial attempt at implementing block support for the Key-value attribute described in #48 by adding another variant of [`NodeBlock`](https://docs.rs/rstml/0.11.2/rstml/node/enum.NodeBlock.html) into [`KeyedAttributeValue`](https://docs.rs/rstml/0.11.2/rstml/node/enum.KeyedAttributeValue.html#). I also implement [`ParseRecoverable`](https://docs.rs/rstml/0.11.2/rstml/recoverable/trait.ParseRecoverable.html#) trait for [`KeyedAttribute`](https://docs.rs/rstml/0.11.2/rstml/node/struct.KeyedAttribute.html#) so that [`transform_block()`](https://docs.rs/rstml/0.11.2/rstml/struct.ParserConfig.html#method.transform_block) will work inside key-value attributes.

```rs
// Placeholder, implement this to meet your requirements
let parse_cfg = ParserConfig::new().transform_block(|input| {
    println!("handled");
    dbg!(input);
    while let Ok(_) = input.parse::<TokenTree>() {}
    return Ok(Some(quote!("handled").into()));
});

let token= quote!(
    <Route
        style={{ width: "10vw" }}
        view={<MyView />}
        ext={anything is possible here!}></Route>
);
dbg!(Parser::new(parse_cfg).parse_simple(token));
```

### CODE REVIEW

File: src/node/attribute.rs [hunk 1]
```diff
@@ -59,10 +59,17 @@ impl AttributeValueExpr {
...
+pub struct AttributeValueBlock {
+    pub token_eq: Token![=],
+    pub value: NodeBlock,
+}
+
 #[derive(Clone, Debug, syn_derive::ToTokens)]
 pub enum KeyedAttributeValue {
     Binding(FnBinding),
     Value(AttributeValueExpr),
+    Block(AttributeValueBlock),
```
Comment (vldm) [inline src/node/attribute.rs:72]: I'd recommend to move of handling invalid block into `AttributeValueExpr`, for two reasons:

1. currently `KeyedAttributeValue` contain completely different paths for parsing
2. currently if macro writer will get `KeyedAttributeValue::Value` he will know that it is value that user typed in `foo=""`

**Review by vldm — COMMENTED** [review]
Hi, thank you for contribution.

Since you anyway plan to use `transform_block` why you need using of `NodeBlock` at all?
The aim of this enum is to save blocks of unparsable rust code aka `InvalidBlock` mainly for IDE purposes (or better error reporting).
It also can be used as replacement for transform_block in some cases.


But if we decide to keep NodeBlock in position of attribute value, i'd recommend to rename this type.

**Comment by tachibanayui** [comment]
> Hi, thank you for contribution.
> 
> Since you anyway plan to use `transform_block` why you need using of `NodeBlock` at all? The aim of this enum is to save blocks of unparsable rust code aka `InvalidBlock` mainly for IDE purposes (or better error reporting). It also can be used as replacement for transform_block in some cases.
> 
> But if we decide to keep NodeBlock in position of attribute value, i'd recommend to rename this type.

From the [rustdoc](https://github.com/rs-tml/rstml/blob/0936bb8ab5ffd59c9f89a9f84f9645884207d8bd/src/node/node_value.rs#L18) of `NodeBlock`, I thought it is a struct to store parsed block or invalid block, and its implementation also handles other cases like `recover_block` so I think it is reasonable to reuse `NodeBlock`. As for renaming, I would suggest `FuzzyBlock`

> I'd recommend to move of handling invalid block into AttributeValueExpr, for two reasons:

I don't get what you mean. Should I implement this inside `AttributeValueExpr::parse_recoverable`?
```rs
// Code copied from NodeBlock::parse_recoverable
Err(e) if parser.config().recover_block => {
    parser.push_diagnostic(e);
    NodeBlock::Invalid(parser.parse_simple(input)?)
}
Err(e) => {
    parser.push_diagnostic(e);
    return None;
}
```

**Comment by vldm** [comment]
@tachibanayui 



 > > I'd recommend to move of handling invalid block into AttributeValueExpr, for two reasons:
 
>I don't get what you mean. Should I implement this inside AttributeValueExpr::parse_recoverable?

You can simply add `transform_block` logic into `ParseRecoverable` implementation of `AttributeValueExpr`. Just do an early check if it is a block, after block transformation you will always get a valid `syn::Block` that can be wrapped into `syn::Expr`.

You don't need to implement `recover_block` logic in the position of attribute value.
We can move this implementation into a separate task along with refining what rust expressions that can be used at the attribute value position.

**Comment by tachibanayui** [comment]
I have updated the implementation to reflect the discussion in #54. 
I also wanted to rename a few structs but hesitated because it would lead to a breaking change:
- `AttributeValueExpr` -> `AttributeValue`
- `KVAttributeValue` -> `AttrValue` (I'm also not sure how to name this)

Demos:
1. Struct initialization syntax (#48), similar to how the `style` prop work in React:
```
<MyComponent style={{ width: "20vw" }} />
```
<details>
  <summary>Parse result: </summary>

```rs
NodeElement {
    open_tag: OpenTag {
        token_lt: Lt,
        name: Path(
            ExprPath {
                attrs: [],
                qself: None,
                path: Path {
                    leading_colon: None,
                    segments: [
                        PathSegment {
                            ident: Ident {
                                sym: MyComponent,
                            },
                            arguments: PathArguments::None,
                        },
                    ],
                },
            },
        ),
        generics: Generics {
            lt_token: None,
            params: [],
            gt_token: None,
            where_clause: None,
        },
        attributes: [
            Attribute(
                KeyedAttribute {
                    key: Path(
                        ExprPath {
                            attrs: [],
                            qself: None,
                            path: Path {
                                leading_colon: None,
                                segments: [
                                    PathSegment {
                                        ident: Ident {
                                            sym: style,
                                        },
                                        arguments: PathArguments::None,
                                    },
                                ],
                            },
                        },
                    ),
                    possible_value: Value(
                        KeyValueAttribute {
                            token_eq: Eq,
                            value: Braced(
                                InvalidBlock {
                                    brace: Brace,
                                    body: TokenStream [
                                        Group {
                                            delimiter: Brace,
                                            stream: TokenStream [
                                                Ident {
                                                    sym: width,
                                                },
                                                Punct {
                                                    char: ':',
                                                    spacing: Alone,
                                                },
                                                Literal {
                                                    lit: "20vw",
                                                    span: bytes(1..7),
                                                },
                                            ],
                                        },
                                    ],
                                },
                            ),
                        },
                    ),
                },
            ),
        ],
        end_tag: OpenTagEnd {
            token_solidus: Some(
                Slash,
            ),
            token_gt: Gt,
        },
    },
    children: [],
    close_tag: None,
},
```

</details>

2. Component in attribute syntax, React Suspend example:
```rs
<Suspend fallback={<div> Loading </div>}>
    <MyView />
</Suspend>
```

<details>
  <summary>Parse result: </summary>

```rs
NodeElement {
    open_tag: OpenTag {
        token_lt: Lt,
        name: Path(
            ExprPath {
                attrs: [],
                qself: None,
                path: Path {
                    leading_colon: None,
                    segments: [
                        PathSegment {
                            ident: Ident {
                                sym: Suspend,
                            },
                            arguments: PathArguments::None,
                        },
                    ],
                },
            },
        ),
        generics: Generics {
            lt_token: None,
            params: [],
            gt_token: None,
            where_clause: None,
        },
        attributes: [
            Attribute(
                KeyedAttribute {
                    key: Path(
                        ExprPath {
                            attrs: [],
                            qself: None,
                            path: Path {
                                leading_colon: None,
                                segments: [
                                    PathSegment {
                                        ident: Ident {
                                            sym: fallback,
                                        },
                                        arguments: PathArguments::None,
                                    },
                                ],
                            },
                        },
                    ),
                    possible_value: Value(
                        AttributeValueExpr {
                            token_eq: Eq,
                            value: Braced(
                                InvalidBlock {
                                    brace: Brace,
                                    body: TokenStream [
                                        Punct {
                                            char: '<',
                                            spacing: Alone,
                                        },
                                        Ident {
                                            sym: div,
                                        },
                                        Punct {
                                            char: '>',
                                            spacing: Alone,
                                        },
                                        Ident {
                                            sym: Loading,
                                        },
                                        Punct {
                                            char: '<',
                                            spacing: Alone,
                                        },
                                        Punct {
                                            char: '/',
                                            spacing: Alone,
                                        },
                                        Ident {
                                            sym: div,
                                        },
                                        Punct {
                                            char: '>',
                                            spacing: Alone,
                                        },
                                    ],
                                },
                            ),
                        },
                    ),
                },
            ),
        ],
        end_tag: OpenTagEnd {
            token_solidus: None,
            token_gt: Gt,
        },
    },
    children: [
        Element(
            NodeElement {
                open_tag: OpenTag {
                    token_lt: Lt,
                    name: Path(
                        ExprPath {
                            attrs: [],
                            qself: None,
                            path: Path {
                                leading_colon: None,
                                segments: [
                                    PathSegment {
                                        ident: Ident {
                                            sym: MyView,
                                        },
                                        arguments: PathArguments::None,
                                    },
                                ],
                            },
                        },
                    ),
                    generics: Generics {
                        lt_token: None,
                        params: [],
                        gt_token: None,
                        where_clause: None,
                    },
                    attributes: [],
                    end_tag: OpenTagEnd {
                        token_solidus: Some(
                            Slash,
                        ),
                        token_gt: Gt,
                    },
                },
                children: [],
                close_tag: None,
            },
        ),
    ],
    close_tag: Some(
        CloseTag {
            start_tag: CloseTagStart {
                token_lt: Lt,
                token_solidus: Slash,
            },
            name: Path(
                ExprPath {
                    attrs: [],
                    qself: None,
                    path: Path {
                        leading_colon: None,
                        segments: [
                            PathSegment {
                                ident: Ident {
                                    sym: Suspend,
                                },
                                arguments: PathArguments::None,
                            },
                        ],
                    },
                },
            ),
            generics: Generics {
                lt_token: None,
                params: [],
                gt_token: None,
                where_clause: None,
            },
            token_gt: Gt,
        },
    ),
}
```
</details>

Let me know if you have any suggestions.

**Comment by vldm** [comment]
@tachibanayui I have added some minor changes directly to your PR, please review, if you approve i am ready to merge it.

**Comment by tachibanayui** [comment]
Hi, I have checked your commit and it's look good to me. Thanks for approving my work!

**Review by vldm — APPROVED** [review]

**Pull request merged**